### ISSUE TITLE
X509_VERIFY_PARAM_add0_policy return value change in 3.1 and 3.0.6 and later

### ISSUE THREAD

**Issue by h-vn** [openssl/openssl#21570]
The manual documents `X509_VERIFY_PARAM_add0_policy()` returning `0` for failure and `1` for success:
https://www.openssl.org/docs/manmaster/man3/X509_VERIFY_PARAM_add0_policy.html

This is no longer true for OpenSSL 3.0.6 and later and for all 3.1 versions. The changed happened between OpenSSL 3.0.5 and 3.0.6 in the following commit:
https://github.com/openssl/openssl/commit/38ebfc3f5f83cbbd01011636d159ad3ed23e9765

While the new return value, count of entries in the stack, makes sense, it's unfortunately now different from the documentation. Note that if only one item is added to the stack, the return value is `1` and those who check for successful operation see no difference. Also any code that checks for failure return code `0` continues to work. Only the code that checks for success return value `1` fails when it adds more than one policy.

Also, a similar change was done for `X509_VERIFY_PARAM_add0_table()` in the same commit. This function is not documented by OpenSSL but OpenBSD [documents](https://man.openbsd.org/X509_VERIFY_PARAM_add0_table.3) it returning `1` or `0` for success and failure, respectively.

Here's the snippet from the commit that showing the change. A further look at the [sk_TYPE_push()](https://www.openssl.org/docs/manmaster/man3/sk_TYPE_push.html) documentation tells that it can return `-1` too, but this seems to not have been considered even before the change below.
```
@@ -339,9 +337,7 @@ int X509_VERIFY_PARAM_add0_policy(X509_VERIFY_PARAM *param,
         if (param->policies == NULL)
             return 0;
     }
-    if (!sk_ASN1_OBJECT_push(param->policies, policy))
-        return 0;
-    return 1;
+    return sk_ASN1_OBJECT_push(param->policies, policy);
 }
 
 int X509_VERIFY_PARAM_set1_policies(X509_VERIFY_PARAM *param,
```

**Comment by mattcaswell** [comment]
That change is definitely a problem and should not have been allowed. It should be reverted IMO.

**Comment by mattcaswell** [comment]
> Here's the snippet from the commit that showing the change. A further look at the [sk_TYPE_push()](https://www.openssl.org/docs/manmaster/man3/sk_TYPE_push.html) documentation tells that it can return -1 too, but this seems to not have been considered even before the change below.

According to the documentation it will only return -1 if `param->policies` is NULL. But that can never be the case here because it is explicitly checked above - so we should never get a -1 response here. Still it might be wise to check anyway just for the sake of robustness.

**Comment by mattcaswell** [comment]
Fix in #21576

**Comment by h-vn** [comment]
> According to the documentation it will only return -1 if param->policies is NULL. But that can never be the case here because it is explicitly checked above - so we should never get a -1 response here.

Oops, I didn't look around enough. Thanks for clarifying this.

And thanks for the fix. The case I noticed this was with TLS certificate Policy OID checking. One use for it has been defined by [eduroam](https://datatracker.ietf.org/doc/html/rfc7593#section-4.2.2) for dynamic RADIUS based WLAN roaming. Another example is EAP-TLS where some organisations use OIDs to further enhance client certificate checking. Custom OID checks may not be very common but they are used, though.
